kolibri/core/decorators.py

```diff
@@ -336,10 +328,20 @@ def cache_no_user_data(view_func):
     on a per user basis.
     """
 
-    @etag(calculate_spa_etag)
-    def inner_func(*args, **kwargs):
+    def calculate_spa_etag(*args, **kwargs):
```

Comment:
My main concern with this approach is that if we are just rendering the entire page anyway, we may as well dispense with the etag and just return the page.

That said, I think that generating the etag from the page contents is definitely the most definitive measure.

At this point, we should take a step back and think about what we are aiming to achieve by the caching of the HTML pages here.

* By caching HTML pages, we reduce server load in two ways:
1. Once a page has been generated, for the next _CACHE_TIMEOUT_ seconds the server can respond with the cached page
2. Once a client has received a page, for the next _CACHE_TIMEOUT_ seconds the client will access the cached copy from its browser cache. When the cache has expired, then an etag can be used. With this solution, the etag no longer saves any server resources.

* We have deliberately set the _CACHE_TIMEOUT_ for these pages to a very low value (15s), this means that most of the value is coming from the server side caching, and the etag. The reason for this low value is to allow for changes in the state used to render the page to propagate quickly to clients (i.e. if the content cache key or theme changes).

We could probably recover some of the benefit of the etag here, by caching the etag value (I know this sounds a little bit crazy) - but this would probably only help if we did a pre-emptive recalculation strategy. i.e. if the etag is requested, but the cache timeout on the etag has not expired yet (but may expire soon) then we would kick off a background recalculation of the etag, rather than do it synchronously.